### PR TITLE
Fix "client-bound" buffers in OpenGL runtime

### PR DESCRIPTION
**Pull request by steven-johnson**
This was pretty much completely broken. HelloAndroidGL now works
properly.

### CODE REVIEW

**Comment by steven-johnson** [comment]
BTW: I think the only code we have in Halide that exercises client-bound GLSL buffers is HelloAndroidGL (and that only does output, not input); should add tests for both somewhere (apps/glsl at a minimum).

File: src/runtime/HalideRuntimeOpenGL.h [hunk 1]
```diff
@@ -34,6 +34,14 @@ extern int halide_opengl_run(void *user_context,
                              int num_coords_dim1);
 // @}
 
+/** When passed to halide_opengl_wrap_texture, this value for texture id is used
+ * to mark the buffer_t.dev field with a constantindicating that the OpenGL object
```
Comment (aam) [inline src/runtime/HalideRuntimeOpenGL.h]: "constant indicating"

File: src/runtime/opengl.cpp [hunk 2]
```diff
@@ -773,6 +772,36 @@ WEAK TextureInfo *find_texture(GLuint tex) {
     return texinfo;
 }
 
+// Allocate new TextureInfo, fill it in, add to global list.
+static TextureInfo *new_texture(GLuint tex, int32_t min[4], int32_t extent[4], bool halide_allocated) {
```
Comment (abadams) [inline src/runtime/opengl.cpp]: Why static instead of weak? Is there a reason to have a separate copy of this per pipeline?

Comment (steven-johnson) [inline src/runtime/opengl.cpp]: ...because I forgot where I was. Fixing.

File: src/runtime/opengl.cpp [hunk 3]
```diff
@@ -228,11 +228,10 @@ WEAK char *strndup(const char *s, size_t n) {
 }
 
 WEAK GLuint get_texture_id(buffer_t *buf) {
-    if (buf->dev == 0) {
-        return 0;
-    } else {
-        return halide_get_device_handle(buf->dev) & 0xffffffff;
-    }
+    uint64_t dev = buf->dev;
```
Comment (zvookin) [inline src/runtime/opengl.cpp]: Does this make the default behavior with a null dev ptr rendering to the frame buffer? Is this intentional and considered a good idea?

Comment (steven-johnson) [inline src/runtime/opengl.cpp]: Ouch, good point. Fixing.

**Comment by steven-johnson** [comment]
PTAL

File: src/runtime/opengl.cpp [hunk 4]
```diff
@@ -788,76 +810,81 @@ WEAK int halide_opengl_device_malloc(void *user_context, buffer_t *buf) {
...
-            return 1;
-        }
+            // Set parameters for this texture: no interpolation and clamp to edges.
+            global_state.BindTexture(GL_TEXTURE_2D, tex);
+            global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+            global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+            global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+            global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+            if (global_state.CheckAndReportError(user_context, "halide_opengl_device_malloc binding texture")) {
+                return 1;
```
Comment (zvookin) [inline src/runtime/opengl.cpp]: This code should call DeleteTextures in the error paths to free the texture names.

**Comment by zvookin** [comment]
The TextureInfo logic seems to prevent binding wrapping a given texture more than once. I think this should be allowed in general, but it absolutely should be allowed for the render target.

**Comment by steven-johnson** [comment]
> The TextureInfo logic seems to prevent binding wrapping a given texture more than once. 
> I think this should be allowed in general, but it absolutely should be allowed for the render target.

I don't necessarily disagree, but that will require some reworking of internal logic; there's currently an assumption that each textureID will correspond to exactly one entry in the TextureInfo list. Can we entertain doing that in a subsequent PR?

**Comment by zvookin** [comment]
Yes, multiple wraps can happen in another PR.

-Z-


On May 5, 2015, at 1:12 PM, Steven Johnson <notifications@github.com> wrote:

The TextureInfo logic seems to prevent binding wrapping a given texture
more than once.
I think this should be allowed in general, but it absolutely should be
allowed for the render target.

I don't necessarily disagree, but that will require some reworking of
internal logic; there's currently an assumption that each textureID will
correspond to exactly one entry in the TextureInfo list. Can we entertain
doing that in a subsequent PR?

—
Reply to this email directly or view it on GitHub
<https://github.com/halide/Halide/pull/775#issuecomment-99207460>.

**Comment by steven-johnson** [comment]
If no objections, gonna land this today.
